### PR TITLE
Chapter 2 solutions

### PR DESCRIPTION
**Pull request by syukronrm**
### Solutions for Chapter 2

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -348,8 +375,10 @@ Implement a function that returns only the first half of a given list.
 >>> firstHalf "bca"
 "b"
 -}
--- PUT THE FUNCTION TYPE IN HERE
-firstHalf l = error "firstHalf: Not implemented!"
+firstHalf :: [a] -> [a]
+firstHalf l = take len l
+  where
+    len = length l `div` 2
```
Comment (vrom911) [inline src/Chapter2.hs:381]: Nice usage of the `where` construction! 👏🏼  Makes code very easily readable 👍🏼

File: src/Chapter2.hs [hunk 2]
```diff
@@ -605,9 +636,18 @@ Implement a function that duplicates each element of the list
...
+-- Solution 1
 duplicate :: [a] -> [a]
-duplicate = error "duplicate: Not implemented!"
+duplicate l = go [] (reverse l)
+  where
+    go :: [a] -> [a] -> [a]
+    go result [] = result
+    go result (x:xs) = go (x : x : result) xs
 
+-- Solution 2, took some time to think about this solution
```
Comment (vrom911) [inline src/Chapter2.hs:647]: Nice one! 

This solution has a number of advantages compared to the previous one: it is more efficient and it is also streaming (the previous one won't work on infinite lists).
Well done 👍🏼

File: src/Chapter2.hs [hunk 3]
```diff
@@ -621,7 +661,10 @@ Write a function that takes elements of a list only in even positions.
 >>> takeEven [2, 1, 3, 5, 4]
 [2,3,4]
 -}
-takeEven = error "takeEven: Not implemented!"
+takeEven :: [a] -> [a]
+takeEven [] = []
+takeEven (x:_:xs) = x : takeEven xs
```
Comment (vrom911) [inline src/Chapter2.hs:666]: Great pattern matching skills 💯

File: src/Chapter2.hs [hunk 4]
```diff
@@ -727,8 +770,10 @@ value of the element itself
 
 🕯 HINT: Use combination of 'map' and 'replicate'
 -}
+-- I don't see where to use `map`
 smartReplicate :: [Int] -> [Int]
-smartReplicate l = error "smartReplicate: Not implemented!"
+smartReplicate [] = []
+smartReplicate (x:xs) = (replicate x x) ++ (smartReplicate xs)
```
Comment (vrom911) [inline src/Chapter2.hs:776]: This is right! 💯 

You can notice that what is happening is you walk through each element of the list, change it to something else (by using `replicate` function) and then concatenate all of the results into one list.

In Haskell, there is the `map` function that does exactly the first part of it (walk through each element of the list, change it to something else ).

After using `map` you will get a list of lists. To flatten such a list, you can use the `concat` function:

```haskell
ghci> concat [[3, 1, 2], [4, 0, 1]]
[3,1,2,4,0,1]
```

This pattern of using `map` and `concat` is very common, so the standard library has even a standard function `concatMap` which is a combination of `map` and `concat` (as the name suggests).

File: src/Chapter2.hs [hunk 5]
```diff
@@ -781,13 +827,14 @@ Let's now try to eta-reduce some of the functions and ensure that we
...
 
--- TODO: type ;)
-listElementsLessThan x l = filter (< x) l
+listElementsLessThan :: Int -> [Int] -> [Int]
+listElementsLessThan x = filter (< x)
 
 -- Can you eta-reduce this one???
-pairMul xs ys = zipWith (*) xs ys
+pairMul :: [Int] -> [Int] -> [Int]
+pairMul = zipWith (*)
```
Comment (vrom911) [inline src/Chapter2.hs:837]: Eta-reduction award :trophy:

File: src/Chapter2.hs [hunk 6]
```diff
@@ -842,7 +889,11 @@ list.
 
 🕯 HINT: Use the 'cycle' function
 -}
-rotate = error "rotate: Not implemented!"
+
+rotate :: Int -> [a] -> [a]
+rotate n l
+  | n < 0 = []
+  | otherwise = take (length l) (drop n (cycle l))
```
Comment (vrom911) [inline src/Chapter2.hs:896]: That's the correct solution! 🌞 

Though, there is a way to improve it a bit 💠  
It does a little bit more than work when the given number is big and the list is small. You will cycle through the list a lot. Maybe you can optimize this using `mod`, for example :slightly_smiling_face:

**Review by vrom911 — APPROVED** [review]
Amazing!
And fantastic work on the advanced task as well 👏🏼

**Comment by chshersh** [comment]
We can't merge solutions to our repository, so I'm closing this PR.
Feel free to open PRs to unsolved chapters in your fork of this repository 🍴

**Pull request closed without merging**